{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "yDneJTaP6UUu"
      },
      "outputs": [],
      "source": [
        "import re\n",
        "!pip install langchain_community\n",
        "from langchain_community.document_loaders import TextLoader\n",
        "from langchain_text_splitters import CharacterTextSplitter"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#attempt at character based splitter, using the line indexes would not work because each chunk has to be assigned to a document\n",
        "\n",
        "#for loop here for each and every one, and then concatenate it to the end\n",
        "#takes a super long time because of how many files there are too split, so may consider just pulling out a few to test on, but we will see that tomorrow\n",
        "text_splitter2 = CharacterTextSplitter(\n",
        "    separator = \"  // \",\n",
        "    chunk_size=100,\n",
        "    chunk_overlap=20,\n",
        "    length_function=len,\n",
        "    is_separator_regex=False,\n",
        ")\n",
        "\n",
        "final_items = []\n",
        "import os\n",
        "directory = r\"/content/drive/MyDrive/biomodels (2)\" #can do all but for the purposes of this, will do a mix of different biomodels across the spectrum (around 30 max!)\n",
        "files = os.listdir(directory)\n",
        "\n",
        "for file in files:\n",
        "    file_path = os.path.join(directory, file)\n",
        "    with open(file_path, 'r') as f:\n",
        "        file_content = f.read()\n",
        "        items = text_splitter2.create_documents([file_content])\n",
        "        final_items.extend(items)"
      ],
      "metadata": {
        "id": "bmGvxKOM6fMh"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "type(items)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "wQvaOGUTr47C",
        "outputId": "f1d9b17b-eb34-473a-f266-93ab766177ad"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "list"
            ]
          },
          "metadata": {},
          "execution_count": 8
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install chromadb\n",
        "!pip install sentence_transformers\n",
        "import chromadb\n",
        "from chromadb.utils import embedding_functions\n",
        "\n",
        "#establishing chroma db\n",
        "CHROMA_DATA_PATH = r\"/content/CHROMA Embeddings\"\n",
        "COLLECTION_NAME = \"BIOMODELS\"\n",
        "EMBED_MODEL = \"all-MiniLM-L6-v2\"\n",
        "client = chromadb.PersistentClient(path = CHROMA_DATA_PATH)\n",
        "\n",
        "embedding_func = embedding_functions.SentenceTransformerEmbeddingFunction(\n",
        "    model_name=EMBED_MODEL\n",
        ")\n",
        "\n",
        "collection = client.create_collection(\n",
        "    name = \"BIOMODELSww\",\n",
        "    embedding_function=embedding_func,\n",
        "    metadata={\"hnsw:space\": \"cosine\"},\n",
        ")"
      ],
      "metadata": {
        "id": "72uQbCSX6tK4"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Uod08CYk7PU9",
        "outputId": "453bb505-3581-4034-af00-61cd5f81f271"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount(\"/content/drive\", force_remount=True).\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import google.generativeai as genai\n",
        "import os\n",
        "\n",
        "os.environ[\"API_KEY\"] = \"AIzaSyBIZakoaE9S0QhbzOw2msHRwczRAl14zmQ\"\n",
        "\n",
        "genai.configure(api_key=os.environ[\"API_KEY\"])\n",
        "\n",
        "model1 = genai.GenerativeModel('gemini-1.5-flash')\n",
        "\n",
        "i=0\n",
        "\n",
        "documents = []\n",
        "\n",
        "for item in final_items:\n",
        "    print(item)\n",
        "    prompt = f'Please summarize this segment of Antimony format code:{item}. The summaries must be clear and concise. For Display Names, provide the value for each variable.  Do not expand mathematical functions into words.'\n",
        "    documents2 = model1.generate_content(prompt)\n",
        "    documents3 = documents2.text\n",
        "    documents.append(documents3) #issue is that each variable is not being defined properly.\n",
        "\n",
        "collection.add(\n",
        "    documents = documents,\n",
        "    ids=[f\"id{i}\" for i in range(len(documents))] #doesnt work because the entire thing is still one whole document\n",
        ")"
      ],
      "metadata": {
        "id": "ajKuWkG97Tit"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "print(documents[7])\n",
        "type(documents)"
      ],
      "metadata": {
        "id": "UfTFJDbI7qnH"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "query_results = collection.query(\n",
        "    query_texts = [\"Give the model that has ATP in it. What is the metadata for this model? \"],\n",
        "    n_results=5,\n",
        ")\n",
        "\n",
        "print(query_results)\n",
        "best_recommendation = query_results[\"documents\"]\n",
        "\n",
        "query_texts = \"Give the model that has ATP in it. What is the metadata for this model? \"\n",
        "\n",
        "prompt_template = prompt_template = f\"\"\"Use the following pieces of context to answer the question at the end. If you don't know the answer, say so.\n",
        "\n",
        "This is the first piece of context necessary: {best_recommendation}\n",
        "\n",
        "Cross-reference all pieces of context to define variables and other unknown entities. Calculate mathematical values based on provided matching variables.\n",
        "\n",
        "Question: {query_texts}\n",
        "\n",
        "\"\"\"\n",
        "response = model1.generate_content(prompt_template)\n",
        "print(response.text)\n",
        "\n",
        "type(response)"
      ],
      "metadata": {
        "id": "IItXkCuv7kSX",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 623
        },
        "outputId": "b2b72dd7-baf5-470f-e309-544c0ffc3b96"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "{'ids': [['id33', 'id25', 'id34', 'id22', 'id0']], 'distances': [[0.4309797941308082, 0.46112877023884713, 0.476316656928498, 0.4781047293455788, 0.48218571449273584]], 'metadatas': [[None, None, None, None, None]], 'embeddings': None, 'documents': [['This segment of Antimony code defines display names for various variables used in the model.\\n\\n**Display Names:**\\n\\n* **substance:** millimole (default)\\n* **time_unit:** hour (default)\\n* **cell:** Erythrocyte\\n* **I:** Ions\\n* **E:** Energy pool\\n* **A:** Adenylate pool\\n* **T:** ATP\\n* **M:** AMP\\n* **P:** Membrane permeability\\n* **J:** Extracellular ion concentration\\n* **W2:** Ion pump activity\\n* **W3:** Glycolytic activity\\n* **U:** de novo AMP synthesis\\n* **W:** AMP degradation\\n* **n:** Dependence of AMP degradation on ATP\\n* **k:** Dependence of AMP degradation on AMP\\n* **U1:** Passive ion influx\\n* **U2:** ATP consumption by ion pump\\n* **U3:** ATP from glycolysis\\n* **U6_plus_3U7:** AMP synthesis de novo\\n* **U6_plus_U7_minus_U8:** Adenylate pool \\n\\n**Metadata:** \\n\\n```\\npage_content=\\'Display Names:\\\\n  substance is \"millimole (default)\";\\\\n  time_unit is \"hour (default)\";\\\\n  cell is \"Erythrocyte\";\\\\n  I is \"Ions\";\\\\n  E is \"Energy pool\";\\\\n  A is \"Adenylate pool\";\\\\n  T is \"ATP\";\\\\n  M is \"AMP\";\\\\n  P is \"Membrane permeability\";\\\\n  J is \"Extracellular ion concentration\";\\\\n  W2 is \"Ion pump activity\";\\\\n  W3 is \"Glycolytic activity\";\\\\n  U is \"de novo AMP synthesis\";\\\\n  W is \"AMP degradation\";\\\\n  n is \"Dependence of AMP degradation on ATP\";\\\\n  k is \"Dependence of AMP degradation on AMP\";\\\\n  U1 is \"Passive ion influx\";\\\\n  U2 is \"ATP consumption by ion pump\";\\\\n  U3 is \"ATP from glycolysis\";\\\\n  U6_plus_3U7 is \"AMP synthesis de novo\";\\\\n  U6_plus_U7_minus_U8 is \"Adenylate pool\";\\'\\n```\\n', 'This segment of Antimony code defines the compartments and species within the model. It specifies that there is one compartment named \"cell\". Within this compartment, there are three species: \"I\", \"E\", and \"A\". \\n\\n**Display Names:** \\n* Compartment: cell\\n* Species: I, E, A\\n\\n**Metadata:** \\n* `page_content=\\'Compartments and Species:\\\\n  compartment cell;\\\\n  species I in cell, E in cell, A in cell;\\'` \\n', '## Summary of Antimony Code Segment:\\n\\nThis Antimony code segment defines the properties of various entities within a model. \\n\\n**Page Content:**\\n\\nThe code defines relationships and properties for different terms, including cell identity, chemical identities (I, E, A), and biological processes (U1, U2, U3, U6_plus_3U7, U6_plus_U7_minus_U8). It provides multiple identifiers (URLs) for each term, likely pointing to different databases.\\n\\n**Display Names:**\\n\\n* **Ataullahkhanov1996_Adenylate:** This is the display name for a specific entity in the model. \\n\\n**Model Entity:**\\n\\n* **Ataullahkhanov1996_Adenylate:** This entity is associated with two BioModels Database entries: MODEL4770526315 and BIOMD0000000054.\\n\\n**Description:**\\n\\n* **Ataullahkhanov1996_Adenylate:** The description for this entity is linked to PubMed ID 8733433.\\n\\n**Taxon:**\\n\\n* **Ataullahkhanov1996_Adenylate:** This entity is associated with the taxonomy ID 9606, representing the species *Homo sapiens* (human).\\n\\n**Part:**\\n\\n* **Ataullahkhanov1996_Adenylate:** This entity is associated with three GO terms: GO:0002028, GO:0006167, and GO:0006110.\\n\\n\\n**Metadata:**\\n\\n* **Format:** Antimony\\n* **Model Name:** Not provided\\n* **Model Version:** Not provided \\n* **Authors:** Not provided\\n* **Creation Date:** Not provided\\n* **Last Modified Date:** Not provided\\n* **Keywords:** Cell Identity, Chemical Identity, Biological Process, Ataullahkhanov1996_Adenylate, BioModels Database, PubMed, Taxonomy, GO Terms \\n', 'This segment of Antimony code defines several terms and properties related to a model of phosphoinositide turnover. These terms include locations like the extracellular space, plasma membrane, cytosol, nucleus, and specific parts of the cell, as well as various molecules like PIP2, IP3, and PLC. It also defines properties like stimulation and activation of specific molecules. The model entity and its associated information are defined as \"Xu2003_Phosphoinositide_turnover\".\\n\\n**Display Names**\\n\\n* page_content = CV terms:\\\\n  Extracellular identity \"http://identifiers.org/go/GO:0005576\"\\\\n  PM identity \"http://identifiers.org/go/GO:0005886\"\\\\n  Cytosol identity \"http://identifiers.org/go/GO:0005622\"\\\\n  NM identity \"http://identifiers.org/go/GO:0031965\"\\\\n  Nucleus identity \"http://identifiers.org/go/GO:0005634\"\\\\n  PIP2_PHGFP_PM part \"http://identifiers.org/chebi/CHEBI:18348\"\\\\n  PIP2_PHGFP_PM part \"http://identifiers.org/interpro/IPR011584\"\\\\n  PIP2_PHGFP_PM part \"http://identifiers.org/interpro/IPR001849\"\\\\n  PH_GFP_Cyt part \"http://identifiers.org/interpro/IPR001849\"\\\\n  PH_GFP_Cyt part \"http://identifiers.org/interpro/IPR011584\"\\\\n  PI_PM hypernym \"http://identifiers.org/chebi/CHEBI:28874\"\\\\n  stim_PM property \"http://identifiers.org/opb/OPB_01072\"\\\\n  IP3_PHGFP_Cyt part \"http://identifiers.org/interpro/IPR001849\"\\\\n  IP3_PHGFP_Cyt part \"http://identifiers.org/chebi/CHEBI:16595\"\\\\n  IP3_PHGFP_Cyt part \"http://identifiers.org/interpro/IPR011584\"\\\\n  PIP2_PM hypernym \"http://identifiers.org/chebi/CHEBI:18348\"\\\\n  PIP_PM hypernym \"http://identifiers.org/chebi/CHEBI:26034\"\\\\n  DAG_PM hypernym \"http://identifiers.org/chebi/CHEBI:18035\"\\\\n  hv_Cytosol hypernym \"http://identifiers.org/chebi/CHEBI:30212\"\\\\n  IP3X_Cytosol identity \"http://identifiers.org/pato/PATO:0002355\"\\\\n  IP3X_Cytosol hypernym \"http://identifiers.org/chebi/CHEBI:16595\"\\\\n  PLC_PM hypernym \"http://identifiers.org/uniprot/Q8K4S1\"\\\\n  PLC_act_PM hypernym \"http://identifiers.org/uniprot/Q8K4S1\"\\\\n  PLC_act_PM property \"http://identifiers.org/ec-code/3.1.4.11\"\\\\n  IP3_Cyt hypernym \"http://identifiers.org/chebi/CHEBI:16595\"\\\\n  PIPSyn hypernym \"http://identifiers.org/biomodels.sbo/SBO:0000216\"\\\\n  PIP2_hyd hypernym \"http://identifiers.org/biomodels.sbo/SBO:0000182\"\\\\n  PLCact hypernym \"http://identifiers.org/biomodels.sbo/SBO:0000170\"\\\\n  PIP2_PH_hyd hypernym \"http://identifiers.org/biomodels.sbo/SBO:0000182\"\\\\n  PIP2_PH hypernym \"http://identifiers.org/biomodels.sbo/SBO:0000182\"\\\\n  IP3deg hypernym \"http://identifiers.org/biomodels.sbo/SBO:0000179\"\\\\n  PIP2Syn hypernym \"http://identifiers.org/biomodels.sbo/SBO:0000216\"\\\\n  IP3_PHGFP hypernym \"http://identifiers.org/biomodels.sbo/SBO:0000182\"\\\\nend\\\\n\\\\nXu2003_Phosphoinositide_turnover is \"Xu2003 - Phosphoinositide turnover\"\\\\n\\\\nXu2003_Phosphoinositide_turnover model_entity_is \"http://identifiers.org/biomodels.db/MODEL3095606944\",\\\\n                                                 \"http://identifiers.org/biomodels.db/BIOMD0000000075\"\\\\nXu2003_Phosphoinositide_turnover identity \"http://identifiers.org/biomodels.db/MODEL3095606944\"\\\\nXu2003_Phosphoinositide_turnover identity \"http://identifiers.org/biomodels.db/BIOMD0000000075\"\\\\nXu2003_Phosphoinositide_turnover parthood \"http://identifiers.org/kegg.pathway/map04070\"\\\\nXu2003_Phosphoinositide_turnover hypernym \"http://identifiers.org/go/GO:0046488\"\\\\nXu2003_Phosphoinositide_turnover container \"http://identifiers.org/bto/BTO:0000931\"\\\\nXu2003_Phosphoinositide_turnover taxon \"http://identifiers.org/taxonomy/10090\"\\\\nXu2003_Phosphoinositide_turnover description \"http://identifiers.org/pubmed/12771127\"\\n\\n**Metadata**\\n\\n* **Format:** Antimony\\n* **Model Entity:** Xu2003_Phosphoinositide_turnover\\n* **Model Name:** Xu2003 - Phosphoinositide turnover\\n* **Model Identifiers:** http://identifiers.org/biomodels.db/MODEL3095606944, http://identifiers.org/biomodels.db/BIOMD0000000075\\n* **KEGG Pathway:** http://identifiers.org/kegg.pathway/map04070\\n* **GO Term:** http://identifiers.org/go/GO:0046488\\n* **BTO Term:** http://identifiers.org/bto/BTO:0000931\\n* **Taxon:** http://identifiers.org/taxonomy/10090\\n* **PubMed ID:** http://identifiers.org/pubmed/12771127 \\n', 'This Antimony code segment defines a model named \"Keizer1996_Ryanodine_receptor_adaptation\". The code snippet utilizes the libAntimony library, version 2.13.0, to generate this model. \\n\\n**Display Names:**\\n\\n* Model: Keizer1996_Ryanodine_receptor_adaptation\\n\\n**Metadata:**\\n\\n* Library: libAntimony\\n* Version: 2.13.0 \\n']], 'uris': None, 'data': None, 'included': ['metadatas', 'documents', 'distances']}\n",
            "The model that has ATP in it is the \"Ataullahkhanov1996_Adenylate\" model. \n",
            "\n",
            "Here's the metadata for the \"Ataullahkhanov1996_Adenylate\" model, derived from the context provided:\n",
            "\n",
            "* **Format:** Antimony\n",
            "* **Model Name:** Not provided \n",
            "* **Model Version:** Not provided \n",
            "* **Authors:** Not provided \n",
            "* **Creation Date:** Not provided \n",
            "* **Last Modified Date:** Not provided \n",
            "* **Keywords:** Cell Identity, Chemical Identity, Biological Process, Ataullahkhanov1996_Adenylate, BioModels Database, PubMed, Taxonomy, GO Terms \n",
            "\n",
            "**Explanation:**\n",
            "\n",
            "1.  We know \"T\" represents ATP from the first piece of context.\n",
            "2.  The third piece of context mentions \"Ataullahkhanov1996_Adenylate\" and associates it with BioModels Database entries. \n",
            "3.  This piece also explains that the entity \"Ataullahkhanov1996_Adenylate\" has a description linked to a PubMed ID, a taxonomy ID, and GO terms.\n",
            "4.  The metadata provided is generic for this type of Antimony code segment, common to all models using this format.\n",
            "\n",
            "Therefore, we can conclude that the \"Ataullahkhanov1996_Adenylate\" model contains ATP, and the metadata provided in the third context is relevant for this model. \n",
            "\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "google.generativeai.types.generation_types.GenerateContentResponse"
            ],
            "text/html": [
              "<div style=\"max-width:800px; border: 1px solid var(--colab-border-color);\"><style>\n",
              "      pre.function-repr-contents {\n",
              "        overflow-x: auto;\n",
              "        padding: 8px 12px;\n",
              "        max-height: 500px;\n",
              "      }\n",
              "\n",
              "      pre.function-repr-contents.function-repr-contents-collapsed {\n",
              "        cursor: pointer;\n",
              "        max-height: 100px;\n",
              "      }\n",
              "    </style>\n",
              "    <pre style=\"white-space: initial; background:\n",
              "         var(--colab-secondary-surface-color); padding: 8px 12px;\n",
              "         border-bottom: 1px solid var(--colab-border-color);\"><b>google.generativeai.types.generation_types.GenerateContentResponse</b><br/>def __init__(done: bool, iterator: None | Iterable[glm.GenerateContentResponse] | AsyncIterable[glm.GenerateContentResponse], result: glm.GenerateContentResponse, chunks: Iterable[glm.GenerateContentResponse] | None=None)</pre><pre class=\"function-repr-contents function-repr-contents-collapsed\" style=\"\"><a class=\"filepath\" style=\"display:none\" href=\"#\">/usr/local/lib/python3.10/dist-packages/google/generativeai/types/generation_types.py</a>Instances of this class manage the response of the `generate_content` method.\n",
              "\n",
              "These are returned by `GenerativeModel.generate_content` and `ChatSession.send_message`.\n",
              "This object is based on the low level `glm.GenerateContentResponse` class which just has `prompt_feedback`\n",
              "and `candidates` attributes. This class adds several quick accessors for common use cases.\n",
              "\n",
              "The same object type is returned for both `stream=True/False`.\n",
              "\n",
              "### Streaming\n",
              "\n",
              "When you pass `stream=True` to `GenerativeModel.generate_content` or `ChatSession.send_message`,\n",
              "iterate over this object to receive chunks of the response:\n",
              "\n",
              "```\n",
              "response = model.generate_content(..., stream=True):\n",
              "for chunk in response:\n",
              "  print(chunk.text)\n",
              "```\n",
              "\n",
              "`GenerateContentResponse.prompt_feedback` is available immediately but\n",
              "`GenerateContentResponse.candidates`, and all the attributes derived from them (`.text`, `.parts`),\n",
              "are only available after the iteration is complete.</pre>\n",
              "      <script>\n",
              "      if (google.colab.kernel.accessAllowed && google.colab.files && google.colab.files.view) {\n",
              "        for (const element of document.querySelectorAll('.filepath')) {\n",
              "          element.style.display = 'block'\n",
              "          element.onclick = (event) => {\n",
              "            event.preventDefault();\n",
              "            event.stopPropagation();\n",
              "            google.colab.files.view(element.textContent, 498);\n",
              "          };\n",
              "        }\n",
              "      }\n",
              "      for (const element of document.querySelectorAll('.function-repr-contents')) {\n",
              "        element.onclick = (event) => {\n",
              "          event.preventDefault();\n",
              "          event.stopPropagation();\n",
              "          element.classList.toggle('function-repr-contents-collapsed');\n",
              "        };\n",
              "      }\n",
              "      </script>\n",
              "      </div>"
            ]
          },
          "metadata": {},
          "execution_count": 17
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "print(best_recommendation)\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "1fOq5wBN3njx",
        "outputId": "d6eaba6f-52f0-473d-def2-bee335525181"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[['This segment of Antimony format text defines initial values for species in a model. It sets the initial concentration of species \"I\" to 10, species \"E\" to 2.1, and species \"A\" to 1.11. These values are likely representing concentrations of molecules or other components in a biological or chemical system. The format \"Species initializations:\" followed by indented lines with species names and their corresponding values is a common way to define initial conditions in Antimony, a modeling language used to describe biochemical networks. \\n', 'This segment of Antimony format text defines the components of a biological system. It sets up three compartments labeled \"cell\" and three species: \"I,\" \"E,\" and \"A.\" Each species is declared to exist within the \"cell\" compartment. This indicates that the model will simulate the interactions and dynamics of these species within a single cellular environment. The format follows a simple syntax, using keywords like \"compartment\" and \"species\" to define the components and their locations.\\n', 'This segment of Antimony format text defines the components and relationships of a model. \\n\\nThe first part, \"page_content\", is defining the terms used throughout the model. It specifies various identifiers for biological entities, such as cell identity, chemical compounds, enzymes, and biological processes. For example, \"cell identity\" is assigned to the Gene Ontology term \"GO:0005623,\" while \"I identity\" is assigned multiple identifiers from different databases, including ChEBI and KEGG.\\n\\nThe next section defines a specific model named \"Ataullahkhanov1996_Adenylate.\" It establishes its identity, model entity, description, taxon, and parts. \\n\\nThe identifier \"Ataullahkhanov1996_Adenylate\" is used to refer to the model itself.  Its \"model_entity_is\" links it to two specific BioModels database entries (MODEL4770526315 and BIOMD0000000054). The model\\'s description references a PubMed article (PMID: 8733433), and its \"taxon\" is assigned to the human species (taxonomy identifier 9606). \\n\\nFinally, the model\\'s \"part\" is defined as a combination of three Gene Ontology terms, \"GO:0002028\", \"GO:0006167\", and \"GO:0006110\", which relate to biological processes such as cell growth and metabolic processes. \\n\\nThis Antimony text essentially creates a structured representation of a biological model, linking its components and relationships to various databases and identifiers, thereby providing a clear and consistent way to define and understand the model\\'s context and function. \\n']]\n"
          ]
        }
      ]
    }
  ]
}